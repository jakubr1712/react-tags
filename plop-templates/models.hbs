export interface I{{properCase name}}ComponentProps {}