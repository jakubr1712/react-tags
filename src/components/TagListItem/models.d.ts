export interface ITagListItemProps {
  tag: CommonTypes.ITag;
  index: number;
  toggleTag: (tag: CommonTypes.ITag) => void;
}
