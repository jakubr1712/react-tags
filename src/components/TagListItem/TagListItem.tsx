import React, { FC } from "react";
import { ITagListItemProps } from "components/TagListItem/models.d";

const TagListItem: FC<ITagListItemProps> = ({ tag, index, toggleTag }) => {
  return (
    <label
      key={tag.id}
      htmlFor={"tag-" + tag.id}
      className={`tag-list__item ${
        index === 0 ? "tag-list__item--highlighted" : ""
      }`}
    >
      <input
        id={"tag-" + tag.id}
        type="checkbox"
        checked={tag.isActive}
        onChange={() => toggleTag(tag)}
        aria-label={tag.name}
      />
      <div>
        <span className="tag-list__name">{tag.name}</span>
        <span className="tag-list__count">+{tag.count}</span>
      </div>
    </label>
  );
};

export default TagListItem;
