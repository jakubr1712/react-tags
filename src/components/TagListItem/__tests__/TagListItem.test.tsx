import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import TagListItem from "../TagListItem";
import { ITagListItemProps } from "../models.d";

describe("TagListItem component", () => {
  const mockToggleTag = jest.fn();

  const defaultProps: ITagListItemProps = {
    tag: { id: 1, name: "Tag1", count: 10, isActive: false },
    index: 0,
    toggleTag: mockToggleTag,
  };

  beforeEach(() => {
    mockToggleTag.mockClear();
  });

  it("renders properly and matches snapshot", () => {
    const { container } = render(<TagListItem {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the correct tag name and count", () => {
    render(<TagListItem {...defaultProps} />);
    expect(screen.getByText("Tag1")).toBeInTheDocument();
    expect(screen.getByText("+10")).toBeInTheDocument();
  });

  it("applies highlighted class to the first tag", () => {
    render(<TagListItem {...defaultProps} />);
    expect(screen.getByLabelText("Tag1").closest("label")).toHaveClass(
      "tag-list__item--highlighted"
    );
  });

  it("calls toggleTag with correct tag when checkbox is clicked", () => {
    render(<TagListItem {...defaultProps} />);
    fireEvent.click(screen.getByLabelText("Tag1"));
    expect(mockToggleTag).toHaveBeenCalledWith(defaultProps.tag);
  });

  it("renders the correct tag name and count for a non-first tag", () => {
    const props = {
      ...defaultProps,
      index: 1,
      tag: { id: 2, name: "Tag2", count: 20, isActive: true },
    };
    render(<TagListItem {...props} />);
    expect(screen.getByText("Tag2")).toBeInTheDocument();
    expect(screen.getByText("+20")).toBeInTheDocument();
    expect(screen.getByLabelText("Tag2").closest("label")).not.toHaveClass(
      "tag-list__item--highlighted"
    );
  });
});
