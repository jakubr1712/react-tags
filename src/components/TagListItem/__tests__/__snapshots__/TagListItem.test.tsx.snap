// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagListItem component renders properly and matches snapshot 1`] = `
<div>
  <label
    class="tag-list__item tag-list__item--highlighted"
    for="tag-1"
  >
    <input
      aria-label="Tag1"
      id="tag-1"
      type="checkbox"
    />
    <div>
      <span
        class="tag-list__name"
      >
        Tag1
      </span>
      <span
        class="tag-list__count"
      >
        +
        10
      </span>
    </div>
  </label>
</div>
`;
